src/ImageSharp/Formats/Jpeg/Port/JpegDecoderCore.cs

```diff
@@ -0,0 +1,968 @@
+                }
+
+                ref FrameComponent component = ref this.frame.Components[componentIndex];
+                int tableSpec = this.InputStream.ReadByte();
+                component.DCHuffmanTableId = tableSpec >> 4;
+                component.ACHuffmanTableId = tableSpec & 15;
+            }
+
+            this.InputStream.Read(this.temp, 0, 3);
+
+            int spectralStart = this.temp[0];
+            int spectralEnd = this.temp[1];
+            int successiveApproximation = this.temp[2];
+            var scanDecoder = default(ScanDecoder);
+
+            scanDecoder.DecodeScan(
+                 this.frame,
+                 this.InputStream,
+                 this.dcHuffmanTables,
+                 this.acHuffmanTables,
+                 this.frame.Components,
+                 componentIndex,
+                 selectorsCount,
+                 this.resetInterval,
+                 spectralStart,
+                 spectralEnd,
+                 successiveApproximation >> 4,
+                 successiveApproximation & 15);
+        }
+
+        /// <summary>
+        /// Build the data for the given component
+        /// </summary>
+        /// <param name="component">The component</param>
+        /// <param name="frameComponent">The frame component</param>
+        private void BuildComponentData(ref Component component, ref FrameComponent frameComponent)
+        {
+            int blocksPerLine = component.BlocksPerLine;
+            int blocksPerColumn = component.BlocksPerColumn;
+            using (var computationBuffer = Buffer<short>.CreateClean(64))
+            using (var multiplicationBuffer = Buffer<short>.CreateClean(64))
+            {
+                Span<short> quantizationTable = this.quantizationTables.Tables.GetRowSpan(frameComponent.QuantizationIdentifier);
+                Span<short> computationBufferSpan = computationBuffer;
+
+                // For AA&N IDCT method, multiplier are equal to quantization
+                // coefficients scaled by scalefactor[row]*scalefactor[col], where
+                //   scalefactor[0] = 1
+                //   scalefactor[k] = cos(k*PI/16) * sqrt(2)    for k=1..7
+                // For integer operation, the multiplier table is to be scaled by 12.
+                Span<short> multiplierSpan = multiplicationBuffer;
+                for (int i = 0; i < 64; i++)
+                {
+                    multiplierSpan[i] = (short)IDCT.Descale(quantizationTable[i] * IDCT.Aanscales[i], 12);
+                }
+
+                for (int blockRow = 0; blockRow < blocksPerColumn; blockRow++)
+                {
+                    for (int blockCol = 0; blockCol < blocksPerLine; blockCol++)
+                    {
+                        int offset = GetBlockBufferOffset(ref component, blockRow, blockCol);
+                        IDCT.QuantizeAndInverseFast(ref frameComponent, offset, ref computationBufferSpan, ref multiplierSpan);
```

Comment:
This method is not fast at all :(
![image](https://user-images.githubusercontent.com/6835152/28068275-157527e6-6645-11e7-823f-3ce8570a33be.png)

I'd rather separate the quantaziation logic from IDCT to make it modular and testable. See no real benefits from doing them in one step. We can optimize later.